pkg/cloudprovider/vsphere/loadbalancer/lbprovider.go

```diff
@@ -0,0 +1,210 @@
+		RetryOnStatuses: nsxtConfig.RetryOnStatusCodes,
+	}
+	cfg := nsxt.Configuration{
+		BasePath:             "/api/v1",
+		Host:                 nsxtConfig.Host,
+		Scheme:               "https",
+		UserAgent:            "nsxt-lb-provider/" + Version,
+		UserName:             nsxtConfig.User,
+		Password:             nsxtConfig.Password,
+		RemoteAuth:           nsxtConfig.RemoteAuth,
+		ClientAuthCertFile:   nsxtConfig.ClientAuthCertFile,
+		ClientAuthKeyFile:    nsxtConfig.ClientAuthKeyFile,
+		CAFile:               nsxtConfig.CAFile,
+		Insecure:             nsxtConfig.InsecureFlag,
+		RetriesConfiguration: retriesConfig,
+	}
+	client, err := nsxt.NewAPIClient(&cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating NSX-T client failed")
+	}
+	return NewNsxtBroker(client), nil
+}
+
+func (p *lbProvider) Initialize(client clientset.Interface, stop <-chan struct{}) {
+	go p.reorg(client.CoreV1().Services(""), stop)
+}
+
+// GetLoadBalancer returns the LoadBalancerStatus
+// Implementations must treat the *corev1.Service parameter as read-only and not modify it.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) GetLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service) (status *corev1.LoadBalancerStatus, exists bool, err error) {
+	servers, err := p.access.FindVirtualServers(clusterName, objectNameFromService(service))
+	if err != nil {
+		return nil, false, err
+	}
+	if len(servers) == 0 {
+		return nil, false, nil
+	}
+	return newLoadBalancerStatus(servers[0].IpAddress), true, nil
+}
+
+func newLoadBalancerStatus(ipAddress string) *corev1.LoadBalancerStatus {
+	return &corev1.LoadBalancerStatus{
+		Ingress: []corev1.LoadBalancerIngress{
+			{IP: ipAddress},
+		},
+	}
+}
+
+// GetLoadBalancerName returns the name of the load balancer. Implementations must treat the
+// *corev1.Service parameter as read-only and not modify it.
+func (p *lbProvider) GetLoadBalancerName(ctx context.Context, clusterName string, service *corev1.Service) string {
+	return clusterName + ":" + service.Namespace + ":" + service.Name
+}
+
+// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
+// Implementations must treat the *corev1.Service and *corev1.Node
+// parameters as read-only and not modify them.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) EnsureLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service, nodes []*corev1.Node) (*corev1.LoadBalancerStatus, error) {
+	key := objectNameFromService(service).String()
+	p.keyLock.Lock(key)
+	defer p.keyLock.Unlock(key)
+
+	class, err := p.classFromService(service)
+	if err != nil {
+		return nil, err
+	}
+
+	state := newState(p.lbService, clusterName, service, nodes)
+	err = state.Process(class)
+	status, err2 := state.Finish()
```

Comment:
nit: this reads a bit counter-intuitive. from what I can tell, this is because `state.Finish` is trying to do two things depending on whether `len(service.spec.Ports) == 0`: cleanup or return a valid status(because `EnsureLoadBalancerDeleted` is leveraging `EnsureLoadBalancer`). Based on the comment and function name, it seems more natural to move the latter part to `Process`, so we can do:
```	
if err != nil {
		return status, err
}
```
right after `status, err = state.Process(class)`

Reply:
Hmm, I think the problems is, that the `Finish` stuff should be done for all returns in `Process`. This could be done with a `defer` here, but this cannot return a new error if it fails. Therefore we separated this into a dedicated method, that can be called with this mixed error code handling.